packages/test/test-end-to-end-tests/src/test/dataMigration.spec.ts

```diff
@@ -0,0 +1,425 @@
+			? context.getSpecifiedCodeDetails()
+			: undefined;
+		assert(codeDetails !== undefined, "get code details failed");
+
+		this._currentRuntimeFactory = codeDetails.package === "v1" ? this.v1 : this.v2;
+		if (this.clientDetailsType === detachedClientType) {
+			const detachedContext = createReadOnlyContext(context, "v2");
+			this._currentRuntimeFactory = this.v1ToV2;
+			return this.currentRuntimeFactory.preInitialize(detachedContext, existing);
+		}
+
+		return this.currentRuntimeFactory.preInitialize(context, existing);
+	}
+
+	public async instantiateFirstTime(runtime: ContainerRuntime): Promise<void> {
+		return this.currentRuntimeFactory.instantiateFirstTime(runtime);
+	}
+
+	public async instantiateFromExisting(runtime: ContainerRuntime): Promise<void> {
+		return this.currentRuntimeFactory.instantiateFromExisting(runtime);
+	}
+
+	public async hasInitialized(runtime: ContainerRuntime): Promise<void> {
+		await this.currentRuntimeFactory.hasInitialized(runtime);
+
+		if (this.clientDetailsType === summarizerClientType) {
+			this.oldRuntime = runtime;
+			const readonlyRuntime = await runtime.loadDetachedAndTransitionFn();
+			this.deferred.resolve({ summarizer: this.oldRuntime, readonlyRuntime });
+		}
+	}
+}
+
+class SharedCellToSharedMapFactory implements IChannelFactory {
+	public static readonly Type = "https://graph.microsoft.com/types/cell";
+
+	public get type(): string {
+		return SharedCellToSharedMapFactory.Type;
+	}
+
+	public get attributes(): IChannelAttributes {
+		return SharedMap.getFactory().attributes;
+	}
+
+	public async load(
+		runtime: IFluidDataStoreRuntime,
+		id: string,
+		services: IChannelServices,
+		attributes: IChannelAttributes,
+	): Promise<SharedMap> {
+		const cell = new SharedCell(id, runtime, attributes);
+		await cell.load(services);
+		const map = SharedMap.getFactory().create(runtime, id) as SharedMap;
+		const val = cell.get();
+		map.set(cellKey, val as string);
+		return map;
+	}
+
+	public create(document: IFluidDataStoreRuntime, id: string): SharedMap {
+		throw new Error("Shouldn't be making a shared cell");
+	}
+}
+
+async function logTree(tree: ISnapshotTree, storage: IDocumentStorageService) {
+	const newTree: any = {};
+	newTree.blobs = {};
+	newTree.trees = {};
+	for (const [blobName, blobId] of Object.entries(tree.blobs)) {
+		const content = await readAndParse(storage, blobId);
+		newTree.blobs[blobName] = content;
+	}
+	for (const [treeName, treeNode] of Object.entries(tree.trees)) {
+		newTree.trees[treeName] = await logTree(treeNode, storage);
+	}
+	// eslint-disable-next-line @typescript-eslint/no-unsafe-return
+	return newTree;
+}
+
+async function migrate(summarizerRuntime: ContainerRuntime, readonlyRuntime: ContainerRuntime) {
+	(readonlyRuntime as any).summarizerNode.startSummary(
+		readonlyRuntime.deltaManager.lastSequenceNumber,
+		new TelemetryNullLogger(),
+	);
+	const appSummary = await readonlyRuntime.summarize({ fullTree: true });
```

Comment:
I'd think you would use createSummary() call instead - I think it's more suitable here.
That said, I think I see the problem with my original thinking - createSummary() will use refSeqNumber = 0, and I believe you are trying to fix it. I wonder if we could generalize it, i.e. can we change our code to be able to create detached containers from arbitrary seq number? I'm sure it's not easy - I'm sure zero is encoded in other places (like driver), though I remember I did some work to remove such code and rely strictly on what Container provides.
I think we need to start collecting notes like that into some readme.md - things that are hard / we have to hack around. As they are negatives for this approach, and having too many of these could outweigh the benefits of this approach over some other approach. I.e. we should always think about it as an experiment, in search for best solution.

That all said, do we care what is the seq number? The app tree should not care, no? Where does it show up?

Reply:
There's some metadata field that I had to modify in the submit summary function on the summarizer client. Look at the containerRuntime submit summary function.